File: src/sa.jl
using Dates
using PyPlot

"""
Make a simulated annealing optimizer.
objective outputs the quality of a solution (Vector -> Float64)
next_temp returns the next temperature to use. The algorithm stops when it is 0.
σ₀ is the starting solution
"""
function make_sa_optimizer(objective::Function, next_temp::Function, neighbor::Function,
    σ₀::Vector)::Function
    T = next_temp()
    σ = σ₀

    function step()
        σ′ = neighbor(σ)
        energy = objective(σ)
        energy′ = objective(σ′)
        curr_energy = energy
        if P(energy, energy′, T) ≥ rand()
            σ = σ′
            curr_energy = energy′
        end
        T = next_temp()

        σ, curr_energy
    end

    step
end

function P(energy, energy′, T)::Float64
    acceptance_prob = if energy′ < energy
        1.0
    else
        exp(-(energy′-energy)/T)
    end

    acceptance_prob
end

"""
Outputs a simple annealing schedule function.
Notice that the first time the return function is called, it will return T₀.
This one seems to work a lot better than the linear schedule.
"""
function make_fast_schedule(T₀::Float64)::Function
    num_steps = -1
    function next_temp()
        num_steps += 1
        T₀ / (num_steps + 1)
    end

    next_temp
end

"""
Make a temperature schedule that decreases linearly. The first temperature returned is
T₀. Each subsequent temperature is δT less than the last. The minimum temperature is 0.
"""
function make_linear_schedule(T₀::Float64, ΔT::Float64)::Function
    T = T₀ + ΔT
    function schedule()::Float64
        T -= ΔT
        max(0.0, T)
    end
    schedule
end

"""
try to create an array where each value is the same as its index
"""
function example_objective(genotype::Vector{Int})::Int
    size = length(genotype)
    fitness = 10*size
    for (i, gene) in enumerate(genotype)
        fitness -= abs(i-gene)
    end
    -fitness
end

function example_neighbor(solution::Vector{Int})::Vector{Int}
    new = copy(solution)
    new[rand(1:length(new))] += rand((1, -1))
    new
end

function sa_demo()
    start_time = Dates.now()
    println("Beginning.")
    sequence_length = 10
    T₀ = 100.0
    max_steps = 500
    σ₀ = [1 for i in 1:sequence_length]
    optimizer_step = make_sa_optimizer(example_objective, make_fast_schedule(T₀),
        example_neighbor, σ₀)
    
    best_solution = missing
    energies = zeros(max_steps)
    for step = 1:max_steps
        best_solution, energy = optimizer_step()
        energies[step] = energy
    end
    println(best_solution)
    println("Done. ($(Dates.now()-start_time))")
    PyPlot.plot(energies)
    PyPlot.show()
end

if abspath(PROGRAM_FILE) == @__FILE__
    sa_demo()
end

File: src/sa.py
from typing import Callable
import numpy as np
import time
import matplotlib.pyplot as plt


def main():
    sa_demo()


def make_sa_optimizer(objective: Callable, next_temp: Callable, neighbor: Callable,
                      sigma0: np.ndarray) -> Callable:
    T = next_temp()
    sigma = sigma0

    def step():
        nonlocal sigma, T
        sigma_prime = neighbor(sigma)
        energy = objective(sigma)
        energy_prime = objective(sigma_prime)
        curr_energy = energy
        if P(energy, energy_prime, T) >= np.random.rand():
            sigma = sigma_prime
            curr_energy = energy_prime
        T = next_temp()

        return sigma, curr_energy

    return step


def P(energy, energy_prime, T) -> float:
    acceptance_prob = 1.0 if energy_prime < energy else np.exp(-(energy_prime-energy)/T)  # type: ignore
    return acceptance_prob


def make_fast_schedule(T0: float) -> Callable:
    num_steps = -1

    def next_temp():
        nonlocal num_steps
        num_steps += 1
        return T0 / (num_steps + 1)

    return next_temp


def make_linear_schedule(T0: float, delta_T: float) -> Callable[[], float]:
    T = T0 + delta_T

    def schedule() -> float:
        nonlocal T
        T -= delta_T
        return max(0, T)

    return schedule


def example_objective(sigma: np.ndarray) -> int:
    size = sigma.shape[0]
    fitness = 10*size
    for i, x in enumerate(sigma):  # type: ignore
        fitness -= abs(i-x)
    return -fitness


def example_neighbor(solution: np.ndarray) -> np.ndarray:
    new = np.copy(solution)
    new[np.random.randint(0, new.shape[0])] += np.random.choice((-1, 1))
    return new


def sa_demo():
    start_time = time.time()
    print('Beginning.')
    sequence_length = 10
    T0 = 100.0
    max_steps = 500
    sigma0 = np.ones(sequence_length, dtype='int')
    optimizer_step = make_sa_optimizer(example_objective, make_fast_schedule(T0),
                                       example_neighbor, sigma0)

    best_solution = None
    energies = np.zeros(max_steps)
    for step in range(max_steps):
        best_solution, energy = optimizer_step()
        energies[step] = energy

    print(best_solution)
    print(f'Done. {time.time()-start_time}')
    plt.plot(energies)
    plt.show()


if __name__ == '__main__':
    main()
